Make metadata lazy for table row classes

benjeffery opened the pull request:
We were decoding metadata on every table row access previously, this makes that lazy. This same method will be applied to the container classes that get returned from e.g. `ts.node(u)` when they are dataclasses.

Note this also gives table row objects a `.table` attribute (@hyanwong might be interested) This attribute does not get put into the result of `dataclasses.asdict` or used in the auto dataclasses methods such as `__eq__` or `__repr__`. These auto methods also only see decoded metadata. The dataclasses are still "frozen".

Needs a couple of extra tests and and one optimisation for the null schema case before merging. Also thinking the schema reference should be stored, as otherwise it could be modified before use by the row.

jeromekelleher commented on python/tskit/tables.py:
```diff
@@ -43,16 +43,55 @@
+    return cls
+
+
+@lazy_metadata
```
Not obvious here why we need both a decorator and to inherit from LazyMetadata - one or other is easy to follow for a reader, but both is confusing. What's the rationale here?

jeromekelleher replied:
I guess we should spell it out up here where these classes are defined anyway, need to encode this knowledge somewhere.

benjeffery replied:
I couldn't figure out how to make an extra slot with a class decorator, annoying!

jeromekelleher replied:
Should we bother with the decorator then? We should be able to implement this in a straighforward inheritancy type way, right?

benjeffery replied:
Manged to do it all in a decorator in the end, can't do inheritance due to intercepting the constructor.

jeromekelleher reviewed (commented):
Looks good, see some comments.

benjeffery commented:
@jeromekelleher I think this could have a review now. Docs are not building due to a warning I haven't figured out yet, but look good ignoring the warning.

Note that I've removed the deprecated attrs on mutation - these could be added back if you think it is too early.

jeromekelleher commented on python/tests/simplify.py:
```diff
@@ -218,7 +219,7 @@ def record_edge(self, left, right, parent, child):
             last = self.edge_buffer[child][-1]
             if last.right == left:
-                last.right = right
+                builtins.object.__setattr__(last, "right", right)
```
This isn't a pattern I want to encourage people to use. This seems worse than not having Frozen instances.

benjeffery replied:
Yes agreed - I could rejig these instances to not use the ugly way. It only really needs to be used in the custom `__init__` and the metadata decorator.

jeromekelleher commented on python/tskit/tables.py:
```diff
@@ -43,54 +42,55 @@
+            and self.metadata == other.metadata
+        )
 
     def __neq__(self, other):
```
Two things here:

1. I think we can get rid of ``__neq__`` now, this is handled automatically in more recent versions of Python
2. Can we use the dataclass ``__eq__``? If not, let's add a comment saying why, as this will be a recurring question.

(Ah, answering my own question here: we have numpy arrays. Good to add a quick comment here to that effect)

jeromekelleher commented on python/tskit/trees.py:
```diff
@@ -78,158 +79,96 @@ def span(self):
+    """
+    metadata: Any
+    """
+    The decoded :ref:`metadata <sec_metadata_definition>`
```
Well, only decoded if a schema is present, right?

jeromekelleher commented on python/tskit/trees.py:
```diff
@@ -304,213 +238,207 @@ def span(self):
+            and np.array_equal(self.mutations, other.mutations)
+        )
+
+    def __neq__(self, other):
```
Don't need neq I think

jeromekelleher commented on python/tskit/trees.py:
```diff
@@ -304,213 +238,207 @@ def span(self):
     """
 
+    # To get default values on a frozen instance you have work around the freeze
     def __init__(
```
Would the post_init hook be better?

benjeffery replied:
Possibly - would still need to use the ugly setattr.

jeromekelleher commented on python/tskit/trees.py:
```diff
@@ -560,71 +488,95 @@ class Variant(SimpleContainer):
     underlying tree sequence data.
 
-    :ivar site: The site object for this variant.
+    :ivar site:
```
We can get rid of the ivars entirely I think (hope?)

jeromekelleher commented on python/tskit/metadata.py:
```diff
@@ -39,6 +41,8 @@
 import tskit
 import tskit.exceptions as exceptions
 
+__setattr__ = builtins.object.__setattr__
```
I wonder about the wisdom of this - don't we then export this into our namespace? If we're going to do something this ugly, the I'd prefer to feel the full brunt of the ugliness each time we use it, just to be reminded of how ugly it is.

benjeffery replied:
I imported it to avoid the double lookup on each row access - I thinking importing it to something called `__builtin__object__setattr__` would be ugly enough?

jeromekelleher reviewed (commented):
Looks good, but I'm seriously questioning whether making things frozen is worth it. What problem are we trying to solve here?

Having to resort to this ``__setattr__`` stuff to do perfectly reasonable things seems much more likely to result in errors than anything the frozen is trying to prevent. Plus, I think we are likely to break code if we change the semantics to frozen now.

If we're trying to allow for things like:
```
tables.nodes[0].flags = 0
```

then we perhaps we could have table rows frozen for now (to allow this in future) but keep the stuff returned by the TreeSequence unfrozen?

jeromekelleher commented:
> Note that I've removed the deprecated attrs on mutation - these could be added back if you think it is too early.

When were they marked as deprecated? We should update the changelog with details.

benjeffery commented:
> Looks good, but I'm seriously questioning whether making things frozen is worth it. What problem are we trying to solve here?

The idea was to prevent users thinking they could mutate the classes - as it warned about in the old docs.  

> Having to resort to this `__setattr__` stuff to do perfectly reasonable things seems much more likely to result in errors than anything the frozen is trying to prevent. Plus, I think we are likely to break code if we change the semantics to frozen now.
> 
> If we're trying to allow for things like:
> 
> ```
> tables.nodes[0].flags = 0
> ```
> 
> then we perhaps we could have table rows frozen for now (to allow this in future) but keep the stuff returned by the TreeSequence unfrozen?

Do you mean the other way round? As in `TableRow` classes are unfrozen to allow for future mutation, and `Node` et al. are frozen? This seems to make sense to me.

jeromekelleher commented:
> Do you mean the other way round? As in TableRow classes are unfrozen to allow for future mutation, and Node et al. are frozen? This seems to make sense to me.

No - to me the only point in freezing the classes is so that we don't break people's code later when we change the semantics to "push the changes back into the tables". So, let people mutate away on the Node, Edge etc classes - we document that these are copies of the original data, and they can do whatever they like with them. Freezing the classes just for the sake of it, or to somehow prevent people from writing bad code seems ... unpythonic to me.

jeromekelleher commented:
How about we make the minimal changes to dataclasses first as cleanly as we can with the same semantics as we currently have (ish), and then look at whether we should freeze or not as a second pass?

benjeffery commented:
@jeromekelleher OK, semantics are now exactly as they were before. Think I have addressed the comments too. Just the docs 
to fix, for some reason it isn't finding the `Site` class.

jeromekelleher commented on python/tests/test_highlevel.py:
```diff
@@ -2991,17 +2971,18 @@ def test_metadata(self):
         (inst2,) = self.get_instances(1)
         assert inst == inst2
-        inst._metadata_decoder = lambda m: "different decoder"
+        inst.metadata
```
What does this statement do?

benjeffery replied:
I thought it was forcing decode of the metadata, but that is done by the equality operator above anyway, removed.

jeromekelleher commented on python/tskit/tables.py:
```diff
@@ -46,51 +45,49 @@
     location: np.ndarray
     parents: np.ndarray
-    metadata: bytes
+    metadata: Any
```
We could say ``metadata : Union[bytes, dict]`` couldn't we?

benjeffery replied:
Changed to `Optional[Union[bytes, dict]]`

jeromekelleher reviewed (approved):
Looks great, seems like a major improvement. One comment on the type annotations.
